### ISSUE TITLE
Lobstermen do not use energy to turn around

### ISSUE THREAD

**Issue by Player701** [OpenXcom/OpenXcom#1249]
**Synopsis:** In vanilla TFTD, Lobstermen aliens expend energy points in addition to time units just when turning around. However, this is not the case in OpenXcom.

Note that I don't have the original version at hand for comparison, but I distinctly remember this particular quirk of the game's toughest alien species. Strangely enough, a quick google search of the OpenXcom forums did not turn up any mentions of this issue. But I'm sure the developers can verify everything first-hand.

**Reproduction steps:** This can be easily reproduced in debug mode, or simply by using MC on enemy units.

**Comment by Player701** [comment]
All right, so I've finally managed to verify the original behavior with the Steam version of TFTD (both DOS and Windows). Here are some facts that I've uncovered about Lobstermen:

- They use 1 energy point for each 45-degree turn, e.g. turning 90 degrees will consume 2 energy
- They _do not_ use additional energy points if they turn while walking. This is consistent with the usual behavior of not expending TUs in the same case.
- They use _twice as much energy_ while walking, e.g. 4 energy points is expended to traverse an ordinary sand tile, while an XCOM soldier needs only 2 points to traverse it.

And now the most important fact of all... **The above seems to hold true for ALL enemy units in TFTD, not just Lobstermen**... but OpenXcom seems to use the standard behavior for all of them.

I haven't been able to verify this for all races so far, only for these ones:

- Aquatoid
- Gillman
- Deep One
- Tasoth
- Lobster Man
- Hallucinoid
- Tentaculat



I am planning to do further research, and will post any updates here. For now, I'm renaming the issue because it's not just about Lobstermen anymore.

**Comment by Player701** [comment]
Confirmed the same behavior for the last three alien races:

- Bio-Drone
- Calcinite
- Triscene

**Civilians also seem to be affected.**

So, it turns out, all non-player units in TFTD indeed use the same logic for energy expenditure, and OpenXcom gets it wrong.

**Comment by Player701** [comment]
I believe I have found the source of the problem. The article on [Energy](https://www.ufopaedia.org/index.php/Energy) at UFOpaedia.org has a formula for energy usage:

> Energy Used = INT( TUs_Used / 2<sup>unitref[45]</sup>)

Looking at my test save files from Steam TFTD, the 45th byte in UNITREF.dat is either 0 or 1, and the number of records where the value is 1 is the same as the number of player units (soldiers + HWPs). All TFTD aliens and civilians seem to have this value set at 0.

OpenXcom does not implement the above formula at all, and just divides the number of TUs spent by 2:

https://github.com/OpenXcom/OpenXcom/blob/430bf7cf47193ed8f104466562b105c14a05355e/src/Savegame/BattleUnit.cpp#L1386-L1399

Energy expenditure when turning is not implemented as well.

**Comment by hellrazor4223** [comment]
Your observation is precise. This will also effect UFO Defense units.
In regard of modding and using bigger maps i am personally unsure if its wise to fully implement this. Aliens not being able to fight back because they run out of energy for moving is kinda bad.

**Comment by Player701** [comment]
Well, I believe the original logic should be used by default, and if someone doesn't want it, they can always use a mod. However, if the developers (or the community as a whole) disagree with me, why not still make this coefficient (equivalent to unitref[45]) moddable? There is a way to make this feature either opt-in or opt-out depending on the general opinion.

I'm currently looking through the code to see how difficult it would be to implement.

And yes, I should probably check the behavior of EU units too. Thank you for reminding me.

**Comment by Player701** [comment]
Confirmed the same behavior for all EU non-player units as well, including civilians.

**Comment by Player701** [comment]
Just stumbled across [this 10-year old thread](http://www.strategycore.co.uk/forums/topic/8407-icing-mod-for-ufo-enemy-unknown/) on StrategyCore. Apparently, the author of that mod thought this issue was a bug in the original game. Well, I don't know... there is way too little information available, so I can't be sure. Or maybe my Google-fu is not strong enough.

Anyway, in either case I think it'd be cool if it were possible to mod this. We should probably get an opinion from the developer team first, though.

**Comment by hellrazor4223** [comment]
Yes indeed the default vanilla behaviour should be used, but there is a big difference in regard of pathfinding. OpenXcom has a pathfinding which actually works UFO and TFTD in the DOS/GOG/Steam Version all suffer from misserable pathing, so enemies will not always move-

**Comment by MeridianOXC** [comment]
Observation confirmed.
To be honest, I believe this is a bug in OG.

**Comment by TBeholder** [comment]
Interesting. Description of  UNITREF.DAT format (linked right there in UFOpaedia) also says that setting this to 4 nullifies all energy expenditure, and that sometimes it’s set to as high as 112 in the original, which may or may not be an error (perhaps it was supposed to be a bitfield, or something?).
Suggestion: perhaps the best solution would be to have a set of “movement cost” profiles, and treat this field as index of the profile to be used? It can be trivially made fully compatible by having the correct defaults (even up to 4, for modding purpose) in standard rulesets. At the same time, this would allow much fine-tuning in mods, for both standard actions and “Alternate movement methods”, with arbitrary unit types.

**Comment by TBeholder** [comment]
… I mean, sure, there’s *moveCost* for armor, which could be expanded to explicitly set all and any TU expenditures that are not covered elsewhere (under Inventory/Item/Skill).  But in any actual ruleset there are going to be much fewer movement modes than Armor anyway. If a mod starts to actually redefine these, it’s likely to become massively redundant and beg for using *refNode* a lot just to keep track of them. So why not keep it tidy, transparent and somewhat more memory/cache -efficient at step #1?
